### PR TITLE
Add CI and Modernize CMake

### PR DESCRIPTION
**Pull request by abeimler**
* Add github workflows
    * windows, ubuntu and macos builds
    * Tests
* Modernize CMake
    * add fixed-containers as a library
    * add `BUILD_TESTS` option

### CODE REVIEW

**Comment by alexkaratarakis** [comment]
Hi @abeimler ,
Thank you for the PR, I am really sorry for the delay on reviewing this.
I will look into it asap.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,30 +1,42 @@
 cmake_minimum_required(VERSION 3.16)
-project(fixed_containers)
+#if(DEFINED ENV{VCPKG_ROOT} AND NOT DEFINED CMAKE_TOOLCHAIN_FILE)
+#  set(CMAKE_TOOLCHAIN_FILE "$ENV{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake" CACHE STRING "")
+#endif()
+#if(NOT DEFINED CMAKE_TOOLCHAIN_FILE)
+#  set(CMAKE_TOOLCHAIN_FILE "${CMAKE_SOURCE_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake" CACHE STRING "")
+#endif()
```
Comment (alexkaratarakis) [inline CMakeLists.txt:7]: Intentionally leaving for merging?
I don't mind leaving this in, but I was curious: do you typically checkout vcpkg inside a repo root?

File: CMakeLists.txt [hunk 2]
```diff
@@ -66,49 +78,76 @@ else()
...
+
+# TODO: split library in components like in bazel
+#add_library(fixed_containers_bidirectional_iterator INTERFACE include/fixed_containers/bidirectional_iterator.hpp)
+#add_library(fixed_containers::bidirectional_iterator ALIAS fixed_containers_bidirectional_iterator)
+#target_include_directories(fixed_containers_bidirectional_iterator INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+
+
+option(BUILD_TESTS "Enable Tests" ${PROJECT_IS_TOP_LEVEL})
+if(BUILD_TESTS)
+    # This variable is set by project() in CMake 3.21+
```
Comment (alexkaratarakis) [inline CMakeLists.txt:102]: Per this comment, PROJECT_IS_TOP_LEVEL is only available in CMake 3.21+ and it is set here for compatibility with lower version, but the variable it used 2 lines above and controls whether this if-block is entered, so this wouldn't work for < 3.21.

The built-in `BUILD_TESTING` might be of use here: https://cmake.org/cmake/help/latest/command/enable_testing.html
According to the docs, it also calls `enable_testing()` automatically.

File: CMakeLists.txt [hunk 3]
```diff
@@ -66,49 +78,76 @@ else()
...
-add_test_dependencies(pair_view_test)
-add_executable(string_literal_test test/string_literal_test.cpp)
-add_test_dependencies(string_literal_test)
-add_executable(type_name_test test/type_name_test.cpp)
-add_test_dependencies(type_name_test)
+add_library(project_warnings INTERFACE)
+target_compile_options(project_warnings INTERFACE ${DIAGNOSTIC_FLAGS})
+
+
+find_package(magic_enum CONFIG)
```
Comment (alexkaratarakis) [inline CMakeLists.txt:85]: Nit: `REQUIRED`

File: CMakeLists.txt [hunk 4]
```diff
@@ -66,49 +78,76 @@ else()
...
+find_package(range-v3 CONFIG)
```
Comment (alexkaratarakis) [inline CMakeLists.txt:86]: `range-v3` is only needed for tests, so it can be placed in the test-only region.

File: .gitignore [hunk 5]
```diff
@@ -41,3 +41,636 @@ user.bazelrc
```
Comment (alexkaratarakis) [inline .gitignore:41]: Is the `.gitignore` change intended (per commit message, it seems like no).
For that matter, the additions are around OS and IDE ignores. Have you considered [Global.gitignore](https://github.com/github/gitignore/tree/main/Global) for those? There are sources about it in the readme in the linked page, but the gist of it is that `Global.gitignore` contains ignores related to a particular user's environment (OS, IDE) and that way you don't have to rely on each repo you work with to have them.

**Comment by alexkaratarakis** [comment]
Also, I will be checking out the CI builds in my fork.

**Comment by alexkaratarakis** [comment]
CI workflow works; I will go ahead and merge this and address the comments after.
Thank you for the contribution @abeimler !

File: CMakeLists.txt (hunk 1 above)
Comment (abeimler) [reply CMakeLists.txt:7]: I just to have this as a git submodule in my projects.

File: .gitignore (hunk 5 above)
Comment (abeimler) [reply .gitignore:41]: Ah sorry, I'm using gitignore.io to generate my gitignore files and uses Clion, MSVC as IDE(s)...

_I will checkout Global.gitignore, thx_

**Pull request merged**